File: app/src/main/java/org/xhanka/biblesiswati/ui/main/VerseDetailsFragment.java
package org.xhanka.biblesiswati.ui.main;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import org.jetbrains.annotations.NotNull;
import org.xhanka.biblesiswati.R;
import org.xhanka.biblesiswati.ui.main.adapter.VerseDetailsAdapter;
import org.xhanka.biblesiswati.ui.main.room.BibleViewModel;

import java.util.Objects;

public class VerseDetailsFragment extends Fragment {


    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        // setSharedElementEnterTransition(TransitionInflater.from(getContext()).inflateTransition(android.R.transition.move));
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        return inflater.inflate(R.layout.fragment_verse_details, container, false);
    }

    @Override
    public void onViewCreated(@NonNull @NotNull View view, @Nullable @org.jetbrains.annotations.Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        RecyclerView recyclerView = view.findViewById(R.id.recyclerView);

        recyclerView.setLayoutManager(new LinearLayoutManager(getContext()));
        recyclerView.addItemDecoration(new DividerItemDecoration(
                view.getContext(),
                DividerItemDecoration.VERTICAL)
        );


        String book_name = getArguments() != null ? getArguments().getString("book_name") : "genesis";
        int chapter_num = getArguments() != null ? getArguments().getInt("chapter_num") : 1;


        BibleViewModel bibleViewModel = new ViewModelProvider(
                this,
                ViewModelProvider.AndroidViewModelFactory.getInstance(
                        Objects.requireNonNull(getActivity()).getApplication())
        ).get(BibleViewModel.class);

        VerseDetailsAdapter adapter = new VerseDetailsAdapter(
                getChildFragmentManager(), bibleViewModel
        );

        recyclerView.setAdapter(adapter);
        bibleViewModel.getBookVerses(book_name, chapter_num).observe(this, adapter::submitList);

    }
}

File: app/src/main/java/org/xhanka/biblesiswati/ui/main/VerseDetailsFragment.kt
package org.xhanka.biblesiswati.ui.main

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.fragment.app.activityViewModels
import androidx.navigation.fragment.navArgs
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.transition.TransitionManager
import dagger.hilt.android.AndroidEntryPoint
import org.xhanka.biblesiswati.R
import org.xhanka.biblesiswati.common.Stagger
import org.xhanka.biblesiswati.ui.main.adapter.VerseDetailsAdapter
import org.xhanka.biblesiswati.ui.main.models.Verse
import org.xhanka.biblesiswati.ui.main.room.BibleViewModel

@AndroidEntryPoint
class VerseDetailsFragment : Fragment() {
    private val stagger = Stagger()
    private val args by navArgs<VerseDetailsFragmentArgs>()

    val bibleViewModel by activityViewModels<BibleViewModel>()

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        return inflater.inflate(R.layout.fragment_verse_details, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        val recyclerView: RecyclerView = view.findViewById(R.id.recyclerView)
        val layoutManager = LinearLayoutManager(context)
        recyclerView.layoutManager = layoutManager
        recyclerView.addItemDecoration(
            DividerItemDecoration(
                view.context,
                DividerItemDecoration.VERTICAL
            )
        )
        val bookName = args.bookName
        val chapterNum = args.chapterNum

        val adapter = VerseDetailsAdapter(childFragmentManager, bibleViewModel)

        // use this variable to display staggering animation on recycler view once,
        // only when the recycler view displays data for the fist time
        var firstTime = true

        // use this variable to scroll to position if user is from favorites fragment
        var shouldScroll = (args.verseNum != -1)

        recyclerView.adapter = adapter
        bibleViewModel.getBookVerses(bookName, chapterNum)
            .observe(viewLifecycleOwner) { verses: List<Verse?> ->
                if (firstTime) {
                    TransitionManager.beginDelayedTransition(recyclerView, stagger)
                    firstTime = false
                }
                adapter.submitList(verses)
                if (shouldScroll) {
                    shouldScroll = false
                    layoutManager.scrollToPosition(args.verseNum - 1)
                }
            }
    }
}